src/liballoc/alloc.rs

```diff
@@ -191,12 +193,89 @@ pub(crate) unsafe fn box_free<T: ?Sized>(ptr: Unique<T>) {
 #[stable(feature = "global_alloc", since = "1.28.0")]
 #[rustc_allocator_nounwind]
 pub fn handle_alloc_error(layout: Layout) -> ! {
-    #[allow(improper_ctypes)]
-    extern "Rust" {
-        #[lang = "oom"]
-        fn oom_impl(layout: Layout) -> !;
+    #[cfg(target_has_atomic = "ptr")]
+    {
+        let hook = HOOK.load(Ordering::SeqCst);
+        let hook: fn(Layout) = if hook.is_null() {
+            default_alloc_error_hook
+        } else {
+            unsafe { mem::transmute(hook) }
+        };
+        hook(layout);
+    }
+    #[cfg(not(target_has_atomic = "ptr"))]
+    {
+        default_alloc_error_hook(layout)
+    }
+
+    unsafe {
+        #[cfg(not(stage0))]
+        {
+            #[allow(improper_ctypes)]
+            extern "Rust" {
+                // This function is generated by the compiler
+                // and calls __rust_abort_internal in src/libstd/alloc.rs
+                // if the `std` crate is linked.
+                fn __rust_maybe_abort_internal();
```

Comment:
Initially reading this I was a little confused by this about why both the hook and the abort mechanism were configurable.  I think, though, it's to do `sys::abort_internal`, right? (as opposed to `intrinsics::abort`)

If that's the case, mind adding a small comment here alluding to that?

Reply:
> it's to do sys::abort_internal, right? (as opposed to intrinsics::abort)

Yes, exactly